controllers/k8ssandra/cleanup.go

```diff
@@ -119,3 +123,193 @@ func (r *K8ssandraClusterReconciler) checkFinalizer(ctx context.Context, kc *api
 
 	return result.Continue()
 }
+
+func (r *K8ssandraClusterReconciler) checkDcDeletion(ctx context.Context, kc *api.K8ssandraCluster, logger logr.Logger) result.ReconcileResult {
+	dcName := k8ssandra.GetDatacenterForDecommission(kc)
+	if dcName == "" {
+		return result.Continue()
+	}
+
+	status := kc.Status.Datacenters[dcName]
+
+	switch kc.Status.Datacenters[dcName].DecommissionProgress {
+	case api.DecommNone:
+		logger.Info("Preparing to updating replication for DC decommission", "DC", dcName)
+		status.DecommissionProgress = api.DecommUpdatingReplication
+		kc.Status.Datacenters[dcName] = status
+		return result.Continue()
+	case api.DecommUpdatingReplication:
+		logger.Info("Waiting for replication updates for DC decommission to complete", "DC", dcName)
+		return result.Continue()
+	default:
+		logger.Info("Proceeding with DC deletion", "DC", dcName)
+		return r.deleteDc(ctx, kc, dcName, logger)
+	}
+}
+
+func (r *K8ssandraClusterReconciler) deleteDc(ctx context.Context, kc *api.K8ssandraCluster, dcName string, logger logr.Logger) result.ReconcileResult {
+	kcKey := utils.GetKey(kc)
+
+	stargate, remoteClient, err := r.findStargateForDeletion(ctx, kcKey, dcName, nil)
+	if err != nil {
+		return result.Error(err)
+	}
+
+	if stargate != nil {
+		if err = remoteClient.Delete(ctx, stargate); err != nil && !errors.IsNotFound(err) {
+			return result.Error(fmt.Errorf("failed to delete Stargate for dc (%s): %v", dcName, err))
+		}
+		logger.Info("Deleted Stargate", "Stargate", utils.GetKey(stargate))
+	}
+
+	reaper, remoteClient, err := r.findReaperForDeletion(ctx, kcKey, dcName, remoteClient)
+	if err != nil {
+		return result.Error(err)
+	}
+
+	if reaper != nil {
+		if err = remoteClient.Delete(ctx, reaper); err != nil && !errors.IsNotFound(err) {
+			return result.Error(fmt.Errorf("failed to delete Reaper for dc (%s): %v", dcName, err))
+		}
+		logger.Info("Deleted Reaper", "Reaper", utils.GetKey(reaper))
+	}
+
+	dc, remoteClient, err := r.findDcForDeletion(ctx, kcKey, dcName, remoteClient)
+	if err != nil {
+		return result.Error(err)
+	}
+
+	if dc != nil {
+		if dc.GetConditionStatus(cassdcapi.DatacenterDecommission) == corev1.ConditionTrue {
+			logger.Info("CassandraDatacenter decommissioning in progress", "CassandraDatacenter", kcKey)
+			// There is no need to requeue here. Reconciliation will be trigger by updates made by cass-operator.
+			return result.Done()
+		}
+
+		if err = remoteClient.Delete(ctx, dc); err != nil && !errors.IsNotFound(err) {
```

Comment:
If Datacenter is deleted, the opt-in annotation should be set first (if we want to decommission). And if we want to decommission, shouldn't this requeue and wait for it to finish decommissioning before you delete it from the Status? It could be part of the cluster for a while until it has finished decommissioning. The instant delete from K8ssandraCluster could give the user indication that the process has finished and they could shutdown the target Kubernetes cluster for example (thus ruining the decommission). Not to mention, it's not guaranteed the decommission is able to start instantly.

Reply:
I'll add the annotation.

As for waiting for the decommission to complete, immediately before the `Delete` there is a check,

```go
		if dc.GetConditionStatus(cassdcapi.DatacenterDecommission) == corev1.ConditionTrue {
			logger.Info("CassandraDatacenter decommissioning in progress", "CassandraDatacenter", kcKey)
			// There is no need to requeue here. Reconciliation will be trigger by updates made by cass-operator.
			return result.Done()
		}

		if err = remoteClient.Delete(ctx, dc); err != nil && !errors.IsNotFound(err) {
			return result.Error(fmt.Errorf("failed to delete CassandraDatacenter (%s): %v", dcName, err))
		}
		logger.Info("Deleted CassandraDatacenter", utils.GetKey(dc))
```
Should I be checking differently?

Reply:
I see it now :) I am missing a requeue.